File: merge_sort.py
def merge_sort(arr):
    if len(arr) <= 1:
        return arr
    # Ділимо масив навпіл
    mid = len(arr) // 2
    left_half = merge_sort(arr[:mid])
    right_half = merge_sort(arr[mid:])
    # Зливаємо дві половини
    return merge(left_half, right_half)

def merge(left, right):
    merged = []
    i = j = 0
    # Порівнюємо елементи і додаємо менші
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
    # Додаємо залишки
    merged.extend(left[i:])
    merged.extend(right[j:])
    
    return merged

def test_sorting_algorithm(sort_func):
    test_arrays = [
        [],                          # порожній
        [5],                        # один елемент
        [3, 1, 4, 1, 5, 9, 2, 6],   # невідсортований
        [1, 2, 3, 4, 5],            # вже відсортований
        [5, 4, 3, 2, 1],            # зворотньо відсортований
        [3, 3, 3, 3],               # однакові елементи
        [-5, 0, 3, -2, 1]           # від'ємні числа
    ]

    for arr in test_arrays:
        print(f"Before: {arr}")
        print(f"After : {merge_sort(arr)}")
        print("---")

test_sorting_algorithm(merge_sort)

File: mergesort.js
function merge_sort(arr){
    if (arr.length <= 1)
        return arr;
    const mid = Math.floor(arr.length / 2);  
    let left = merge_sort(arr.slice(0, mid)) 
    let right = merge_sort(arr.slice(mid, arr.length)) 
    //Зливаємо дві половини
    return merge(left, right)  
}
function merge(left, right){
    let merged = [];
    let i = 0, j = 0;
    //Порівнюємо елементи і додаємо менші
    while (i < left.length && j < right.length){
        if(left[i] < right[j]){
            merged.push(left[i])
            i++
        }
        else{
            merged.push(right[j])
            j++
        }
    }
    // Додаємо залишки
    while (i < left.length) {
        merged.push(left[i]);
        i++;
    }
    while (j < right.length) {
        merged.push(right[j]);
        j++;
    }
    return merged;
}

console.log(merge_sort([38, 27, 43, 3, 9, 82, 10]));